### PR TITLE
Issue619 c2d flow (happy path)

### LINKED ISSUE #619: C2D Flow
- [x] Implement `test_compute_registered_dataset` (#630)
- Fix TODOs introduced in 630. See for details: https://github.com/oceanprotocol/ocean.py/pull/630#issue-1093777094
    - [x] Reinstate type enforcement on `DataServiceProvider.start_compute_job` - Unable to fix, so left disabled
    - [x] Fix docstring for `DataServiceProvider.start_compute_job` (what is algorithm_custom_data)?
    - [x] Refactor compute flow tests to use high-level `OceanCompute` and `OceanAssets` APIs instead of `DataServiceProvider` and `ERC20Token` APIs.
    - [x] Pass valid compute environment ids to `initialize` and `startCompute` endpoints - Won't Fix until Provider supports compute environment
- [x] Reinstate `test_compute_raw_algo`
- [x] Reinstate `test_compute_multi_inputs`
- [x] Reinstate `test_update_trusted_algorithms` (blocked by https://github.com/oceanprotocol/ocean.py/pull/646)
- [x] Reinstate `test_compute_trusted_algorithms`
- [x] Review `test_ocean_compute`.  Tests are currently passing, but we suspect updates are required to bring them up to date for v4

### PR DESCRIPTION
**Pull request by DMats**
Towards #619

This PR does not include all of the desired c2d integration tests but it's getting large and we want to review/merge it before implementing the others.

Changes proposed in this PR:

- Implement the happy path for the v4 compute flow integration tests.  The test is named `test_compute_registered_algo` and most of the logic is implemented in `run_compute_test`, a helper meant to be reused for all the compute flow tests.
- Skip more complex compute flows (raw algorithms, multiple dataset inputs, trusted algorithms, etc.)
- Removed unnecessary `Setup` class from `test_compute_flow.py`. Use pytest fixtures instead.
- Add execute permission to `artifacts/install-remote.sh`
- Add `Asset.get_service_by_index` and `Asset.get_service_index_by_id`
- Use `Asset.get_service_index_by_id` whenever calling `ERC20Token.start_order` or `DataServiceProvider.start_compute_job`
- Fix signatures in `DataServiceProvider` (added consumer address to the signed message)
- Temporarily disable `@enforce_types` on `DataServiceProvider.start_compute_job`
- Update input arguments to `DataServiceProvider.start_compute_job` 
- Fix `test_start_compute_job_fails_empty` and `test_start_compute_job_fails_error_response`
- Remove old test helpers `get_algorithm_meta`, `get_registered_ddo`.
- Move `create_basics` and `create_asset` to `ddo_helpers.py`
- Increase default cap to 100 when using `deploy_erc721_erc20`
- Create `AssetAndUserdata` dataclass for use in compute flow tests
- Refer to assets as "asset" instead of "ddo"

# Next Steps

This PR contains 17 skipped tests that should be reinstated.
This PR adds several TODOs that still need to be fixed.
  - Reinstate type enforcement on `DataServiceProvider.start_compute_job`
  - Fix docstring for `DataServiceProvider.start_compute_job` (what is algorithm_custom_data)?
  - Refactor compute flow tests to use high-level `OceanCompute` and `OceanAssets` APIs instead of `DataServiceProvider` and `ERC20Token` APIs.
  - Pass valid compute environment ids to `initialize` and `startCompute` endpoints once the provider actually supports compute environments

### CODE REVIEW

File: tests/integration/test_compute_flow.py [hunk 1]
```diff
@@ -6,211 +6,153 @@
...
+    # Order asset compute service
+    order_tx_id, service = process_order(
         ocean_instance,
         publisher_wallet,
         consumer_wallet,
-        compute_ddo,
+        asset_ddo,
         ServiceTypes.CLOUD_COMPUTE,
     )
-    compute_inputs = [ComputeInput(did, order_tx_id, service.index, userdata=userdata)]
```
Comment (calina-c) [inline tests/integration/test_compute_flow.py:144]: please reinstate this as plural, we need to be able to process multiple inputs.

File: tests/resources/ddo_helpers.py [hunk 2]
```diff
@@ -148,50 +141,94 @@ def get_registered_ddo_with_access_service(ocean_instance, wallet, provider_uri=
...
+                "entrypoint": "node $ALGO",
+                "image": "node",
+                "tag": "10",
+                "checksum": "test",
+            },
+        }
+    }
+    metadata.update(algorithm_values)
+
+    access_service = Service(
```
Comment (calina-c) [inline tests/resources/ddo_helpers.py:214]: you can use the option that autodefines the access service, implicitly. By default, if there are no services given, we create the access service. And there's also options to auto-deploy ERCs. They are much simpler to use and you would require less boilerplate here, in compute. I'd like to see as little code as possible related to publishing here, and focus on compute stuff.

In test_ocean_assets.py you can see examples of publishing both regular assets and compute assets, which you can freely reuse. So the publish part just needs to be transfered here, and used accordingly.

**Review by calina-c — CHANGES_REQUESTED** [review]

File: tests/integration/test_compute_flow.py (hunk 1 above)
Comment (DMats) [reply tests/integration/test_compute_flow.py:144]: I intend to do this once the basic compute flow is working.

File: tests/resources/ddo_helpers.py (hunk 2 above)
Comment (DMats) [reply tests/resources/ddo_helpers.py:214]: >you can use the option that autodefines the access service, implicitly. By default, if there are no services given, we create the access service. And there's also options to auto-deploy ERCs. 
@calina-c Are you referring to `test_ocean_assets.create_asset()`?

Comment (calina-c) [reply tests/resources/ddo_helpers.py:214]: yes

**Comment by calina-c** [comment]
@DMats take care! Your data service provider tests are failing. I'll add skips to the compute tests on the v4main branch and from now on we'll stick to green CI tests.

File: ocean_lib/assets/asset.py [hunk 3]
```diff
@@ -163,16 +163,41 @@ def add_service(self, service: Service) -> None:
...
     def get_service(self, service_type: str) -> Optional[Service]:
-        """Return the first Service with the given service type."""
+        """Return first Service with the given service type.
+        Return None if service with the given service type not found."""
         return next(
             (service for service in self.services if service.type == service_type), None
         )
 
+    @enforce_types
+    def get_service_index_by_id(self, service_id: str) -> int:
```
Comment (calina-c) [inline ocean_lib/assets/asset.py:189]: I like this helper, but I really don't like this function's name. I would suggest the parameter to be a service, not the service id, and let's name it something more suggestive. E.g. "get_index_of_service"?

This is very good, and it helped me find an error within the consume flows.

File: ocean_lib/data_provider/data_service_provider.py [hunk 4]
```diff
@@ -234,60 +236,45 @@ def sign_message(wallet: Wallet, msg: str) -> Tuple[str, str]:
         return nonce, sign_hash(encode_defunct(text=f"{msg}{nonce}"), wallet)
 
     @staticmethod
-    @enforce_types
+    # TODO reinstate @enforce_types
+    # @enforce_types
```
Comment (calina-c) [inline ocean_lib/data_provider/data_service_provider.py:240]: I think you can already handle this TODO. Or do you want to leave it for the second step?

File: ocean_lib/data_provider/data_service_provider.py [hunk 5]
```diff
@@ -655,24 +646,19 @@ def _get_file_name(response: Response) -> Optional[str]:
             return None
 
     @staticmethod
-    @enforce_types
+    # TODO: reinstate @enforce_types
+    # @enforce_types
```
Comment (calina-c) [inline ocean_lib/data_provider/data_service_provider.py:650]: same question here

File: tests/resources/ddo_helpers.py [hunk 6]
```diff
@@ -92,106 +90,184 @@ def get_access_service(
...
+            "created": "2020-11-15T12:27:48Z",
+            "updated": "2021-05-17T21:58:02Z",
+            "description": "Sample description",
+            "name": "Sample asset",
+            "type": "dataset",
+            "author": "OPF",
+            "license": "https://market.oceanprotocol.com/terms",
+        }
+    data_provider = DataServiceProvider
+    file1_dict = {"type": "url", "url": "https://url.com/file1.csv", "method": "GET"}
```
Comment (calina-c) [inline tests/resources/ddo_helpers.py:119]: I added the encrypt function on assets in my other branch (marketplace flow). If you think it helps here with the files, you can add it too

**Review by calina-c — CHANGES_REQUESTED** [review]
Just a few requested changes. Otherwise this is exemplary work!

File: ocean_lib/assets/asset.py (hunk 3 above)
Comment (calina-c) [reply ocean_lib/assets/asset.py:189]: The function exists on the marketplace-flow branch now, so you can copy it over.

File: tests/resources/ddo_helpers.py (hunk 6 above)
Comment (DMats) [reply tests/resources/ddo_helpers.py:119]: Done.

File: ocean_lib/data_provider/data_service_provider.py (hunk 4 above)
Comment (DMats) [reply ocean_lib/data_provider/data_service_provider.py:240]: I tried another quick attempt to get this working but was unsuccessful.  I'll take another look at it later.

**Review by calina-c — APPROVED** [review]

**Pull request merged**